activerecord/lib/active_record/connection_handling.rb

```diff
@@ -304,15 +335,24 @@ def with_role(role, prevent_writes, &blk)
         end
       end
 
-      def with_shard(shard, role, prevent_writes)
-        old_shard = current_shard
+      def with_role_and_shard(role, shard, prevent_writes)
+        prevent_writes = true if role == reading_role
 
-        with_role(role, prevent_writes) do
-          self.current_shard = shard
-          yield
+        if ActiveRecord::Base.legacy_connection_handling
+          with_role(role, prevent_writes) do
+            self.role_and_shard_stack << { role: role, shard: shard, klass: self }
+            yield
+          end
+        else
+          connection_handler.while_preventing_writes(prevent_writes) do
```

Comment:
There's a bug here. We need to move the write prevention into the shard stack, otherwise we can switch connections but the writes prevention remains global. Alternatively it could be moved back to the connection level.